src/hotspot/share/runtime/thread.cpp

```diff
@@ -436,36 +436,60 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
+bool Thread::is_JavaThread_protected(const JavaThread* p) {
   Thread* current_thread = Thread::current();
-  if (!checkTLHOnly) {
-    // Do the simplest check first:
-    if (SafepointSynchronize::is_at_safepoint()) {
-      // The target is protected since JavaThreads cannot exit
-      // while we're at a safepoint.
-      return true;
-    }
 
-    // If the target hasn't been started yet then it is trivially
-    // "protected". We assume the caller is the thread that will do
-    // the starting.
-    if (p->osthread() == NULL || p->osthread()->get_state() <= INITIALIZED) {
-      return true;
-    }
+  // Do the simplest check first:
+  if (SafepointSynchronize::is_at_safepoint()) {
+    // The target is protected since JavaThreads cannot exit
+    // while we're at a safepoint.
+    return true;
+  }
 
-    // Now make the simple checks based on who the caller is:
-    if (current_thread == p || Threads_lock->owner() == current_thread) {
-      // Target JavaThread is self or calling thread owns the Threads_lock.
-      // Second check is the same as Threads_lock->owner_is_self(),
-      // but we already have the current thread so check directly.
-      return true;
-    }
+  // If the target hasn't been started yet then it is trivially
+  // "protected". We assume the caller is the thread that will do
+  // the starting.
+  if (p->osthread() == NULL || p->osthread()->get_state() <= INITIALIZED) {
+    return true;
+  }
+
+  // Now make the simple checks based on who the caller is:
+  if (current_thread == p || Threads_lock->owner() == current_thread) {
+    // Target JavaThread is self or calling thread owns the Threads_lock.
+    // Second check is the same as Threads_lock->owner_is_self(),
+    // but we already have the current thread so check directly.
+    return true;
+  }
+
+  // Check the ThreadsLists associated with the calling thread (if any)
+  // to see if one of them protects the target JavaThread:
+  if (is_JavaThread_protected_by_TLH(p)) {
+    return true;
   }
 
+  // Use this debug code with -XX:+UseNewCode to diagnose locations that
+  // are missing a ThreadsListHandle or other protection mechanism:
+  // guarantee(!UseNewCode, "current_thread=" INTPTR_FORMAT " is not protecting p="
+  //           INTPTR_FORMAT, p2i(current_thread), p2i(p));
+
+  // Note: Since 'p' isn't protected by a TLH, the call to
+  // p->is_handshake_safe_for() may crash, but we have debug bits so
+  // we'll be able to figure out what protection mechanism is missing.
+  assert(p->is_handshake_safe_for(current_thread), "JavaThread=" INTPTR_FORMAT
+         " is not protected and not handshake safe.", p2i(p));
+
+  // The target JavaThread is not protected so it is not safe to query:
+  return false;
+}
+
+// Is the target JavaThread protected by a ThreadsListHandle (TLH) associated
+// with the calling Thread?
+//
+bool Thread::is_JavaThread_protected_by_TLH(const JavaThread* p) {
```

Comment:
I think you could pass JavaThread* current as the first parameter here.  The naming will help to keep it straight and make this an assert(current == JavaThread::current() if this is deemed to be expensive.
At the callsites below, it would make it obvious that "this" isn't meant to be the current thread (and calling Thread::current() inside a (tbd) assert would make it clear which thread is which.

With this handshake code I continuously am trying to figure out which thread is current vs. target.

Reply:
I'm not fond of passing the current thread in as a parameter because that could
make folks think that these functions can be called on any `Thread*`. They must
be called in the context of the current thread which is also why they are static.
I would like to rename the `p` parameter to `target` and I could add `/*target*/`
to the call sites to make things more clear.